File: leetcode/77-combinations/main.py
class Solution(object):
    """
    Implementation similar to permutations

    select k from n number, it is a combination problem
    therefore, the time complexity is nCk

    nCk = n!/(n-k)!k!

    Time    O(nCk)
    Space   O(nCk) due to the recrusion
    beats   13.88%
    """

    def __init__(self):
        self.result = []

    def combine(self, n, k):
        if k < 1 or k > n:
            return []
        # construct the arrary from 1 to n
        nums = []
        for i in range(n):
            nums.append(i+1)
        # dfs
        self.dfs(nums, [], k)
        return self.result

    def dfs(self, nums, path, k):
        if len(path) == k:
            self.result.append(path)
        elif len(path) < k:
            for i in range(len(nums)):
                self.dfs(nums[i+1:], path + [nums[i]], k)


s = Solution()

r = s.combine(4, 2)
print(r)
r = s.combine(4, 3)
print(r)
r = s.combine(5, 4)
print(r)


class Solution(object):
    """
    Implementation similar to permutations
    Time    O(nCk)
    Space   O(nCk) due to the recrusion
    648 ms, faster than 39.65%
    """
    def combine(self, n, k):
        self.res = []
        self.dfs(n, 1, k, [])
        return self.res
    
    def dfs(self, n, start, k, chosen):
        if len(chosen) == k:
            self.res.append(chosen)
            return
        for i in range(start, n+1):
            self.dfs(n, i+1, k, chosen + [i])


s = Solution()

r = s.combine(4, 2)
print(r)
r = s.combine(4, 3)
print(r)
r = s.combine(5, 4)
print(r)

File: leetcode/77-combinations/main.go
package main

import "fmt"

/*
	Recursive implementation similar to permutations
	Time	O(nCk)
	Space O(nCk) due to the recrusion
	beats 1.22% (python version beats 13.88%)
*/
func combine(n int, k int) [][]int {
	if k < 1 || k > n {
		return [][]int{}
	}
	result := [][]int{}
	var dfs func(nums []int, path []int)
	dfs = func(nums []int, path []int) {
		if len(path) == k {
			result = append(result, path)
		} else {
			for i := 0; i < len(nums); i++ {
				// copy thing
				nextpath := []int{}
				nextpath = append(nextpath, path...)
				nextpath = append(nextpath, nums[i])
				// dfs(nums[i+1:], path+[nums[i]])
				dfs(nums[i+1:], nextpath)
			}
		}
	}
	arr := []int{}
	for i := 0; i < n; i++ {
		arr = append(arr, i+1)
	}
	dfs(arr, []int{})
	return result
}

/*
	Recursive implementation similar to permutations(without inner functions)
	Time	O(nCk)
	Space O(nCk) due to the recrusion
	beats 9.76% better due to the inner function removal
*/
func combine1(n int, k int) [][]int {
	if k < 1 || k > n {
		return [][]int{}
	}
	return dfs(1, n, []int{}, k)
}

func dfs(start int, n int, path []int, k int) [][]int {
	result := [][]int{}
	if len(path) == k {
		return [][]int{path}
	} else {
		for i := start; i <= n; i++ {
			// copy thing
			nextpath := []int{}
			nextpath = append(nextpath, path...)
			nextpath = append(nextpath, i)
			// dfs(nums[i+1:], path+[nums[i]])
			temp := dfs(i+1, n, nextpath, k)
			result = append(result, temp...)
		}
	}
	return result
}

/*
	Iterative using a stack
	Time	O(nCk)
	Space O(nCk) the stack
	beats 12.20%
*/

func combine2(n int, k int) [][]int {
	if k < 1 || k > n {
		return [][]int{}
	}
	result := [][]int{}
	stack := []Stack{}
	stack = append(stack, Stack{1, []int{}})
	for len(stack) > 0 {
		pop := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if len(pop.Path) == k {
			result = append(result, pop.Path)
		} else {
			for i := pop.Start; i <= n; i++ {
				nextpath := []int{}
				nextpath = append(nextpath, pop.Path...)
				nextpath = append(nextpath, i)
				stack = append(stack, Stack{i + 1, nextpath})
			}
		}
	}
	return result
}

type Stack struct {
	Start int
	Path  []int
}

func main() {
	fmt.Println(combine(4, 2))
	fmt.Println(combine(4, 3))
	fmt.Println(combine(5, 4))

	fmt.Println(combine2(4, 2))
	fmt.Println(combine2(4, 3))
	fmt.Println(combine2(5, 4))
}
